### PR TITLE
WordSeg: use Tensors everywhere

### PR DESCRIPTION
**Pull request by sgugger**
Test using Tensor of floats everywhere to avoid copying back to the CPU needlessly.
This leads to an AD failure (or I did something stupid) since the model does not produce any gradients anymore. Any help welcome :)

### CODE REVIEW

File: Models/Text/WordSeg/Lattice.swift [hunk 1]
```diff
@@ -232,7 +232,7 @@ extension Lattice.Node {
 
 extension Lattice.Edge {
   public func isAlmostEqual(to other: Self, tolerance: Float) -> Bool {
-    let diffP = abs(self.logp - other.logp)
+    let diffP = abs(self.logp - other.logp).scalarized()
```
Comment (texasmichelle) [inline Models/Text/WordSeg/Lattice.swift:235]: This line is only necessary because `logp` was previously `Float`. Now that it's `Tensor<Float>`, the check in the return statement can use `Tensor.isAlmostEqual`, similar to `self.score` and `self.totalScore`.

File: Models/Text/WordSeg/SemiRing.swift [hunk 2]
```diff
@@ -105,8 +97,8 @@ extension SemiRing {
   // TODO(abdulras) see if we can use ulp as a default tolerance
   @inlinable
   public func isAlmostEqual(to other: Self, tolerance: Float) -> Bool {
-    let diffP = abs(self.logp - other.logp)
-    let diffR = abs(self.logp - other.logp)
+    let diffP = abs(self.logp - other.logp).scalarized()
+    let diffR = abs(self.logp - other.logp).scalarized()
```
Comment (texasmichelle) [inline Models/Text/WordSeg/SemiRing.swift:101]: Similarly, these can also use `Tensor.isAlmostEqual`

**Comment by sgugger** [comment]
I don't have the ability to run the swift-format used in google3, so unless the instructions to build it have become easier than last time I checked, it would be easier if you could run it through the branch.

(I thought swift-format was now auto-triggered?)

File: Models/Text/WordSeg/Lattice.swift (hunk 1 above)
Comment (sgugger) [reply Models/Text/WordSeg/Lattice.swift]: I did the quick fix for the merge conflict but this makes more sense, thanks!

File: Models/Text/WordSeg/SemiRing.swift (hunk 2 above)
Comment (sgugger) [reply Models/Text/WordSeg/SemiRing.swift]: Same as above.

File: Examples/WordSeg/main.swift [hunk 3]
```diff
@@ -72,15 +72,15 @@ for epoch in 1...maxEpochs {
...
+    let (loss, gradients) = valueWithGradient(at: model) { model -> Tensor<Float> in
       let lattice = model.buildLattice(sentence, maxLen: maxLength)
       let score = lattice[sentence.count].semiringScore
       let expectedLength = exp(score.logr - score.logp)
       let loss = -1 * score.logp + lambd * expectedLength
       return loss
     }
 
-    trainingLossSum += loss
+    trainingLossSum += loss.scalarized()
```
Comment (BradLarson) [inline Examples/WordSeg/main.swift:83]: We may want to have the training and validation loss sums be expressed as Tensor<Float> to avoid the `scalarized()` usage here, to make this a little easier on X10. It's a small thing compared to the other challenges this presents to X10, but it's a quick fix.

File: Models/Text/WordSeg/Lattice.swift [hunk 4]
```diff
@@ -54,13 +54,13 @@ public struct Lattice: Differentiable {
         SemiRing(
           logp: logp,
           // TODO(abdulras): this should really use integeral pow
-          logr: logp + logf(powf(Float(sentence.count), Float(order))))
+          logr: logp + Tensor(logf(powf(Float(sentence.count), Float(order)))))
```
Comment (BradLarson) [inline Models/Text/WordSeg/Lattice.swift:57]: This isn't the only case of this, but to make sure of device consistency in creation of new tensors, we should probably add `on: logp.device` so that this is at least the same device as the incoming Tensor argument. Otherwise, it will always be on the default TF eager device.

File: Models/Text/WordSeg/Model.swift [hunk 5]
```diff
@@ -197,10 +187,9 @@ public struct SNLM: EuclideanDifferentiable, KeyPathIterable {
   }
 
   // MARK: - buildLattice
-
-  func get_logp_lex(_ logp_lex: [Float], _ candidate: CharacterSequence) -> Float {
+  func get_logp_lex(_ logp_lex: Tensor<Float>, _ candidate: CharacterSequence) -> Tensor<Float> {
     guard let index = parameters.strVocab.dictionary[candidate] else {
-      return -Float.infinity
+      return Tensor(-Float.infinity)
```
Comment (BradLarson) [inline Models/Text/WordSeg/Model.swift:192]: Same as above, maybe specifying `on: logp_lex.device` would aid in compatibility.

**Comment by texasmichelle** [comment]
I ran lint and it came back clean 👍

**Comment by texasmichelle** [comment]
When #486 is merged, linting should be automatic 🎉

**Comment by sgugger** [comment]
Eh those random failures are annoying in the sense they are not always giving the same error message! Thanks for unblocking me :)

**Comment by sgugger** [comment]
As discussed, I'll add @BradLarson optimizations in a separate PR.

**Comment by texasmichelle** [comment]
Yeah the tests have been picking up old images - in this case, the failure had nothing to do with your changes. Thank you for these updates!
